File: content/develop/use-cases/feature-store/go/feature_store.go
// Package featurestore is a Redis online feature store backed by per-entity
// Hashes.
//
// Each entity (here, a user) lives at a deterministic key such as
// "fs:user:{id}". The hash holds every feature for that entity as one
// field per feature — batch-materialized aggregates (refreshed on a
// daily cycle) alongside streaming-updated signals (refreshed every
// few seconds). One HMGET returns whichever subset the model needs in
// one network round trip.
//
// Two TTL layers solve the *mixed staleness* problem:
//
//   - A key-level EXPIRE aligned with the batch materialization cycle
//     causes the whole entity to disappear if its batch refresher
//     fails, so inference sees a missing entity (which the model
//     handler can detect and fall back on) rather than silently
//     outdated values.
//   - A per-field HEXPIRE on each streaming field gives that field
//     its own shorter expiry, independent of the rest of the hash.
//     When the streaming pipeline stops updating a field, the field
//     self-cleans while the rest of the entity stays populated.
//
// HEXPIRE and HTTL require Redis 7.4 or later. The go-redis v9 client
// exposes them as HExpire and HTTL on *redis.Client.
//
// Concurrency is by construction: Redis is single-threaded per shard,
// so overlapping HSET calls from a batch job and a streaming worker
// on the same entity hash are applied atomically without locks or
// version columns.
package featurestore

import (
	"context"
	"errors"
	"fmt"
	"sort"
	"strconv"
	"sync/atomic"
	"time"

	"github.com/redis/go-redis/v9"
)

// FeatureValue is the concrete type a single feature may take before
// it gets serialized as a Redis hash field. Hash field values are
// strings on the wire; the helper renders these types into strings
// via encode() so booleans round-trip cleanly through redis-cli.
type FeatureValue any

// FeatureMap is the set of fields written for one entity.
type FeatureMap map[string]FeatureValue

// DefaultBatchFields is the schema bulk-loaded once per batch cycle.
var DefaultBatchFields = []string{
	"country_iso",
	"risk_segment",
	"account_age_days",
	"tx_count_7d",
	"avg_amount_30d",
	"chargeback_count_180d",
}

// DefaultStreamingFields is the schema updated by the streaming worker
// with a per-field HEXPIRE so each field self-expires when its
// upstream pipeline stops.
var DefaultStreamingFields = []string{
	"last_login_ts",
	"last_device_id",
	"tx_count_5m",
	"failed_logins_15m",
	"session_country",
}

// Stats holds the helper's in-process counters. Read with FeatureStore.Stats.
type Stats struct {
	BatchWritesTotal     int64 `json:"batch_writes_total"`
	StreamingWritesTotal int64 `json:"streaming_writes_total"`
	ReadsTotal           int64 `json:"reads_total"`
	ReadFieldsTotal      int64 `json:"read_fields_total"`
}

// FeatureStore wraps a *redis.Client and exposes the four feature-store
// paths: batch ingest (BulkLoad), streaming ingest (UpdateStreaming),
// inference read (GetFeatures), and batch scoring (BatchGetFeatures).
type FeatureStore struct {
	rdb                 *redis.Client
	KeyPrefix           string
	BatchTTL            time.Duration
	StreamingTTL        time.Duration

	batchWritesTotal     atomic.Int64
	streamingWritesTotal atomic.Int64
	readsTotal           atomic.Int64
	readFieldsTotal      atomic.Int64
}

// NewFeatureStore returns a FeatureStore backed by rdb. Defaults match
// the Python and Node.js demos: a 24-hour key-level TTL and a 5-minute
// per-field streaming TTL.
func NewFeatureStore(rdb *redis.Client, keyPrefix string, batchTTL, streamingTTL time.Duration) *FeatureStore {
	if keyPrefix == "" {
		keyPrefix = "fs:user:"
	}
	if batchTTL == 0 {
		batchTTL = 24 * time.Hour
	}
	if streamingTTL == 0 {
		streamingTTL = 5 * time.Minute
	}
	return &FeatureStore{
		rdb:          rdb,
		KeyPrefix:    keyPrefix,
		BatchTTL:     batchTTL,
		StreamingTTL: streamingTTL,
	}
}

// KeyFor returns the Redis key for an entity ID.
func (fs *FeatureStore) KeyFor(entityID string) string {
	return fs.KeyPrefix + entityID
}

// -------------------------------------------------------------------
// Batch ingestion (materialization)
// -------------------------------------------------------------------

// BulkLoad materializes a batch of entities into Redis. rows is
// keyed by entity ID. One HSET plus one EXPIRE per entity, batched
// through go-redis's Pipeline so the whole batch ships in a single
// round trip. The key-level EXPIRE is what makes the entity
// disappear if a future batch run fails — inference reads the
// missing entity rather than silently outdated values.
func (fs *FeatureStore) BulkLoad(ctx context.Context, rows map[string]FeatureMap, ttl time.Duration) (int, error) {
	if ttl == 0 {
		ttl = fs.BatchTTL
	}
	if len(rows) == 0 {
		return 0, nil
	}

	pipe := fs.rdb.Pipeline()
	for entityID, fields := range rows {
		key := fs.KeyFor(entityID)
		encoded := make(map[string]any, len(fields))
		for name, value := range fields {
			encoded[name] = encode(value)
		}
		pipe.HSet(ctx, key, encoded)
		pipe.Expire(ctx, key, ttl)
	}
	if _, err := pipe.Exec(ctx); err != nil {
		return 0, fmt.Errorf("bulk load: %w", err)
	}
	fs.batchWritesTotal.Add(int64(len(rows)))
	return len(rows), nil
}

// UpdateBatchFeature overwrites one batch feature without touching
// the key TTL. Used by the demo's "manually refresh one user" lever;
// real pipelines flow through BulkLoad.
func (fs *FeatureStore) UpdateBatchFeature(ctx context.Context, entityID, field string, value FeatureValue) error {
	if err := fs.rdb.HSet(ctx, fs.KeyFor(entityID), field, encode(value)).Err(); err != nil {
		return err
	}
	fs.batchWritesTotal.Add(1)
	return nil
}

// -------------------------------------------------------------------
// Streaming ingestion
// -------------------------------------------------------------------

// UpdateStreaming writes streaming features with a per-field TTL.
//
// Each field carries its own HEXPIRE so it self-expires
// independently of the rest of the hash. If the streaming pipeline
// stops, the streaming fields drop out while the batch-materialized
// fields remain populated under their longer key-level EXPIRE.
//
// HEXPIRE returns one status code per field:
//
//   - 1: TTL set / updated
//   - 2: the expiry was 0 or in the past, so Redis deleted the field
//     instead of applying a TTL
//   - 0: an NX | XX | GT | LT conditional flag was specified and not
//     met (we never use one here)
//   - -2: no such field, or no such key
//
// Since we just HSET every field on the same call, any code other
// than 1 means the per-field TTL invariant did not hold — the
// mixed-staleness story relies on every streaming field carrying a
// fresh TTL after the write, so failure is loud.
func (fs *FeatureStore) UpdateStreaming(ctx context.Context, entityID string, fields FeatureMap, ttl time.Duration) error {
	if len(fields) == 0 {
		return nil
	}
	if ttl == 0 {
		ttl = fs.StreamingTTL
	}
	key := fs.KeyFor(entityID)
	encoded := make(map[string]any, len(fields))
	names := make([]string, 0, len(fields))
	for name, value := range fields {
		encoded[name] = encode(value)
		names = append(names, name)
	}

	pipe := fs.rdb.Pipeline()
	pipe.HSet(ctx, key, encoded)
	hexpireCmd := pipe.HExpire(ctx, key, ttl, names...)
	if _, err := pipe.Exec(ctx); err != nil {
		return fmt.Errorf("update streaming: %w", err)
	}
	codes, err := hexpireCmd.Result()
	if err != nil {
		return fmt.Errorf("update streaming: HEXPIRE: %w", err)
	}
	for _, code := range codes {
		if code != 1 {
			return fmt.Errorf("HEXPIRE did not set every field TTL for %s: %v", key, codes)
		}
	}
	fs.streamingWritesTotal.Add(int64(len(fields)))
	return nil
}

// -------------------------------------------------------------------
// Inference reads
// -------------------------------------------------------------------

// GetFeatures returns a subset of features for one entity. Pass
// fieldNames=nil to fetch the full hash with HGETALL — useful for
// debugging but rarely the right call on the request path, where the
// model knows exactly which features it consumes.
func (fs *FeatureStore) GetFeatures(ctx context.Context, entityID string, fieldNames []string) (map[string]string, error) {
	key := fs.KeyFor(entityID)
	if fieldNames == nil {
		out, err := fs.rdb.HGetAll(ctx, key).Result()
		if err != nil {
			return nil, err
		}
		fs.readsTotal.Add(1)
		fs.readFieldsTotal.Add(int64(len(out)))
		return out, nil
	}
	if len(fieldNames) == 0 {
		return map[string]string{}, nil
	}
	values, err := fs.rdb.HMGet(ctx, key, fieldNames...).Result()
	if err != nil {
		return nil, err
	}
	out := make(map[string]string, len(fieldNames))
	for i, name := range fieldNames {
		if values[i] == nil {
			continue
		}
		s, ok := values[i].(string)
		if !ok {
			continue
		}
		out[name] = s
	}
	fs.readsTotal.Add(1)
	fs.readFieldsTotal.Add(int64(len(out)))
	return out, nil
}

// BatchGetFeatures pipelines HMGET across many entities for batch
// scoring. Returns one map per entity ID, in input order.
func (fs *FeatureStore) BatchGetFeatures(ctx context.Context, entityIDs, fieldNames []string) (map[string]map[string]string, error) {
	if len(entityIDs) == 0 || len(fieldNames) == 0 {
		return map[string]map[string]string{}, nil
	}

	pipe := fs.rdb.Pipeline()
	cmds := make([]*redis.SliceCmd, len(entityIDs))
	for i, id := range entityIDs {
		cmds[i] = pipe.HMGet(ctx, fs.KeyFor(id), fieldNames...)
	}
	if _, err := pipe.Exec(ctx); err != nil && !errors.Is(err, redis.Nil) {
		return nil, fmt.Errorf("batch get features: %w", err)
	}

	out := make(map[string]map[string]string, len(entityIDs))
	var seen int64
	for i, id := range entityIDs {
		values, err := cmds[i].Result()
		if err != nil {
			return nil, fmt.Errorf("batch get features: %s: %w", id, err)
		}
		row := make(map[string]string, len(fieldNames))
		for j, name := range fieldNames {
			if values[j] == nil {
				continue
			}
			if s, ok := values[j].(string); ok {
				row[name] = s
				seen++
			}
		}
		out[id] = row
	}
	fs.readsTotal.Add(int64(len(entityIDs)))
	fs.readFieldsTotal.Add(seen)
	return out, nil
}

// -------------------------------------------------------------------
// TTL inspection (used by the demo UI)
// -------------------------------------------------------------------

// KeyTTLSeconds returns the seconds until the entity key expires:
// positive means TTL remaining, -1 means no key-level TTL set,
// -2 means the key doesn't exist.
func (fs *FeatureStore) KeyTTLSeconds(ctx context.Context, entityID string) (int64, error) {
	d, err := fs.rdb.TTL(ctx, fs.KeyFor(entityID)).Result()
	if err != nil {
		return 0, err
	}
	// go-redis returns time.Duration(-1) for "no TTL" and
	// time.Duration(-2) for "missing key" (both literal nanosecond
	// values, not seconds). Positive durations carry the real TTL.
	if d < 0 {
		return int64(d), nil
	}
	return int64(d.Seconds()), nil
}

// FieldTTLsSeconds returns the per-field TTL for each named field
// via HTTL. Each value mirrors the TTL convention: positive means
// seconds remaining, -1 means the field has no TTL set, -2 means
// the field doesn't exist on this hash (or the key itself is
// missing).
func (fs *FeatureStore) FieldTTLsSeconds(ctx context.Context, entityID string, fieldNames []string) (map[string]int64, error) {
	if len(fieldNames) == 0 {
		return map[string]int64{}, nil
	}
	codes, err := fs.rdb.HTTL(ctx, fs.KeyFor(entityID), fieldNames...).Result()
	if err != nil {
		return nil, err
	}
	// HTTL on a missing key returns an array of -2s, one per field, so
	// the loop below produces the same shape as a present-but-empty
	// hash would. No defensive shim needed for this client.
	out := make(map[string]int64, len(fieldNames))
	for i, name := range fieldNames {
		if i < len(codes) {
			out[name] = codes[i]
		} else {
			out[name] = -2
		}
	}
	return out, nil
}

// -------------------------------------------------------------------
// Demo housekeeping
// -------------------------------------------------------------------

// ListEntityIDs returns up to limit entity IDs by scanning
// keyPrefix*. SCAN is non-blocking and is used to populate UI
// dropdowns, not as a serving primitive. The result is sorted.
func (fs *FeatureStore) ListEntityIDs(ctx context.Context, limit int64) ([]string, error) {
	if limit <= 0 {
		limit = 200
	}
	pattern := fs.KeyPrefix + "*"
	prefixLen := len(fs.KeyPrefix)
	ids := make([]string, 0, limit)
	iter := fs.rdb.Scan(ctx, 0, pattern, 200).Iterator()
	for iter.Next(ctx) {
		k := iter.Val()
		if len(k) <= prefixLen {
			continue
		}
		ids = append(ids, k[prefixLen:])
		if int64(len(ids)) >= limit {
			break
		}
	}
	if err := iter.Err(); err != nil {
		return nil, err
	}
	sort.Strings(ids)
	return ids, nil
}

// CountEntities returns the true count of entities under the key
// prefix. Iterates SCAN without an in-memory cap so the UI can report
// the real total even when more keys exist than the dropdown lists.
func (fs *FeatureStore) CountEntities(ctx context.Context) (int64, error) {
	var n int64
	pattern := fs.KeyPrefix + "*"
	iter := fs.rdb.Scan(ctx, 0, pattern, 500).Iterator()
	for iter.Next(ctx) {
		n++
	}
	if err := iter.Err(); err != nil {
		return 0, err
	}
	return n, nil
}

// DeleteEntity drops one entity by ID. Returns 1 if a key was
// deleted, 0 otherwise.
func (fs *FeatureStore) DeleteEntity(ctx context.Context, entityID string) (int64, error) {
	return fs.rdb.Del(ctx, fs.KeyFor(entityID)).Result()
}

// Reset drops every entity under the key prefix. Used by the demo
// reset path. Scans in batches and issues one variadic DEL per batch,
// so a large demo dataset doesn't land on the server as one giant
// synchronous delete.
func (fs *FeatureStore) Reset(ctx context.Context) (int64, error) {
	var deleted int64
	pattern := fs.KeyPrefix + "*"
	batch := make([]string, 0, 500)
	flush := func() error {
		if len(batch) == 0 {
			return nil
		}
		n, err := fs.rdb.Del(ctx, batch...).Result()
		if err != nil {
			return err
		}
		deleted += n
		batch = batch[:0]
		return nil
	}
	iter := fs.rdb.Scan(ctx, 0, pattern, 500).Iterator()
	for iter.Next(ctx) {
		batch = append(batch, iter.Val())
		if len(batch) >= 500 {
			if err := flush(); err != nil {
				return deleted, err
			}
		}
	}
	if err := iter.Err(); err != nil {
		return deleted, err
	}
	if err := flush(); err != nil {
		return deleted, err
	}
	return deleted, nil
}

// Stats returns a snapshot of the in-process counters.
func (fs *FeatureStore) Stats() Stats {
	return Stats{
		BatchWritesTotal:     fs.batchWritesTotal.Load(),
		StreamingWritesTotal: fs.streamingWritesTotal.Load(),
		ReadsTotal:           fs.readsTotal.Load(),
		ReadFieldsTotal:      fs.readFieldsTotal.Load(),
	}
}

// ResetStats zeroes every counter.
func (fs *FeatureStore) ResetStats() {
	fs.batchWritesTotal.Store(0)
	fs.streamingWritesTotal.Store(0)
	fs.readsTotal.Store(0)
	fs.readFieldsTotal.Store(0)
}

// encode renders a feature value as a string for hash storage.
// Booleans become "true" / "false" so they round-trip cleanly through
// other clients and redis-cli.
func encode(value FeatureValue) string {
	switch v := value.(type) {
	case nil:
		return ""
	case string:
		return v
	case bool:
		if v {
			return "true"
		}
		return "false"
	case int:
		return strconv.FormatInt(int64(v), 10)
	case int32:
		return strconv.FormatInt(int64(v), 10)
	case int64:
		return strconv.FormatInt(v, 10)
	case float32:
		return strconv.FormatFloat(float64(v), 'f', -1, 32)
	case float64:
		return strconv.FormatFloat(v, 'f', -1, 64)
	default:
		return fmt.Sprintf("%v", v)
	}
}

File: content/develop/use-cases/feature-store/nodejs/featureStore.js
"use strict";

/**
 * Redis online feature store backed by per-entity Hashes.
 *
 * Each entity (here, a user) lives at a deterministic key such as
 * `fs:user:{id}`. The hash holds every feature for that entity as one
 * field per feature — batch-materialized aggregates (refreshed on a
 * daily cycle) alongside streaming-updated signals (refreshed every
 * few seconds). One `HMGET` returns whichever subset the model needs
 * in one network round trip.
 *
 * Two TTL layers solve the *mixed staleness* problem:
 *
 *   - A key-level `EXPIRE` aligned with the batch materialization
 *     cycle causes the whole entity to disappear if its batch
 *     refresher fails, so inference sees a missing entity (which the
 *     model handler can detect and fall back on) rather than silently
 *     outdated values.
 *   - A per-field `HEXPIRE` on each streaming field gives that field
 *     its own shorter expiry, independent of the rest of the hash.
 *     When the streaming pipeline stops updating a field, the field
 *     self-cleans while the rest of the entity stays populated.
 *
 * `HEXPIRE` and `HTTL` require Redis 7.4 or later. node-redis 5
 * exposes them as `hExpire` and `hTTL`.
 *
 * Concurrency is by construction: Redis is single-threaded per shard,
 * so overlapping `HSET` calls from a batch job and a streaming worker
 * on the same entity hash are applied atomically without locks or
 * version columns.
 */

/**
 * @typedef {string|number|boolean} FeatureValue
 * @typedef {Record<string, FeatureValue>} FeatureMap
 */

/**
 * Default batch feature schema. Daily aggregates computed offline and
 * bulk-loaded once per materialization cycle.
 */
const DEFAULT_BATCH_FIELDS = Object.freeze([
  "country_iso",
  "risk_segment",
  "account_age_days",
  "tx_count_7d",
  "avg_amount_30d",
  "chargeback_count_180d",
]);

/**
 * Default streaming feature schema. Updated by the streaming worker as
 * new events arrive, with a per-field TTL so each field self-expires
 * when its upstream pipeline stops.
 */
const DEFAULT_STREAMING_FIELDS = Object.freeze([
  "last_login_ts",
  "last_device_id",
  "tx_count_5m",
  "failed_logins_15m",
  "session_country",
]);

/**
 * Encode a feature value as a string for hash storage.
 *
 * Booleans become `"true"` / `"false"` (not `"True"` / `"False"`) so
 * they round-trip cleanly through other clients and `redis-cli`.
 *
 * @param {FeatureValue} value
 * @returns {string}
 */
function encode(value) {
  if (typeof value === "boolean") return value ? "true" : "false";
  return String(value);
}

class FeatureStore {
  /**
   * @param {object} options
   * @param {import("redis").RedisClientType} options.redisClient
   * @param {string} [options.keyPrefix="fs:user:"]
   * @param {number} [options.batchTtlSeconds=86400]
   * @param {number} [options.streamingTtlSeconds=300]
   */
  constructor({
    redisClient,
    keyPrefix = "fs:user:",
    batchTtlSeconds = 24 * 60 * 60,
    streamingTtlSeconds = 5 * 60,
  } = {}) {
    if (!redisClient) {
      throw new Error("A connected redisClient is required.");
    }
    this.redis = redisClient;
    this.keyPrefix = keyPrefix;
    this.batchTtlSeconds = batchTtlSeconds;
    this.streamingTtlSeconds = streamingTtlSeconds;

    // Node.js is single-threaded for JS execution, so plain numbers
    // are safe for counters. No lock needed.
    this.batchWritesTotal = 0;
    this.streamingWritesTotal = 0;
    this.readsTotal = 0;
    this.readFieldsTotal = 0;
  }

  // --- Key helpers ---------------------------------------------------

  /** @param {string} entityId */
  keyFor(entityId) {
    return `${this.keyPrefix}${entityId}`;
  }

  // --- Batch ingestion (materialization) -----------------------------

  /**
   * Materialize a batch of entities into Redis.
   *
   * `rows` is `{entityId: {field: value, ...}}`. One `HSET` plus one
   * `EXPIRE` per entity, all batched into a single round trip through
   * `multi().exec()`. The key-level `EXPIRE` is what makes the whole
   * entity disappear if a future batch run fails — inference reads
   * the missing entity rather than silently outdated values.
   *
   * @param {Record<string, FeatureMap>} rows
   * @param {number} [ttlSeconds]
   * @returns {Promise<number>}
   */
  async bulkLoad(rows, ttlSeconds) {
    const ttl = ttlSeconds ?? this.batchTtlSeconds;
    const ids = Object.keys(rows);
    if (ids.length === 0) return 0;

    const pipe = this.redis.multi();
    for (const entityId of ids) {
      const key = this.keyFor(entityId);
      const fields = rows[entityId];
      const encoded = {};
      for (const [name, value] of Object.entries(fields)) {
        encoded[name] = encode(value);
      }
      pipe.hSet(key, encoded);
      pipe.expire(key, ttl);
    }
    await pipe.exec();
    this.batchWritesTotal += ids.length;
    return ids.length;
  }

  /**
   * Update a single batch feature without touching the key TTL.
   *
   * Used by the demo's "manually refresh one user" lever; in a real
   * pipeline batch updates always flow through `bulkLoad`.
   *
   * @param {string} entityId
   * @param {string} field
   * @param {FeatureValue} value
   * @returns {Promise<void>}
   */
  async updateBatchFeature(entityId, field, value) {
    await this.redis.hSet(this.keyFor(entityId), field, encode(value));
    this.batchWritesTotal += 1;
  }

  // --- Streaming ingestion -------------------------------------------

  /**
   * Write streaming features with a per-field TTL.
   *
   * Each field carries its own `HEXPIRE` so it self-expires
   * independently of the rest of the hash. If the streaming pipeline
   * stops, the streaming fields drop out while the batch-materialized
   * fields remain populated under their longer key-level `EXPIRE`.
   *
   * `HEXPIRE` returns one status code per field:
   *   1 = TTL set / updated,
   *   2 = the expiry was 0 or in the past, so Redis deleted the field
   *       instead of applying a TTL,
   *   0 = an `NX | XX | GT | LT` conditional flag was specified and not
   *       met (we never use one here),
   *   -2 = no such field, or no such key.
   * We just `HSET` every field on the same call, so any code other
   * than 1 means the per-field TTL invariant did not hold — the
   * mixed-staleness story relies on every streaming field carrying a
   * fresh TTL after the write, so failure is loud.
   *
   * @param {string} entityId
   * @param {FeatureMap} fields
   * @param {number} [ttlSeconds]
   * @returns {Promise<void>}
   */
  async updateStreaming(entityId, fields, ttlSeconds) {
    const names = Object.keys(fields);
    if (names.length === 0) return;
    const ttl = ttlSeconds ?? this.streamingTtlSeconds;
    const key = this.keyFor(entityId);
    const encoded = {};
    for (const [name, value] of Object.entries(fields)) {
      encoded[name] = encode(value);
    }

    const [, expireResult] = await this.redis
      .multi()
      .hSet(key, encoded)
      .hExpire(key, names, ttl)
      .exec();
    if (!Array.isArray(expireResult) ||
        expireResult.some((code) => Number(code) !== 1)) {
      throw new Error(
        `HEXPIRE did not set every field TTL for ${key}: ` +
          JSON.stringify(expireResult),
      );
    }
    this.streamingWritesTotal += names.length;
  }

  // --- Inference reads -----------------------------------------------

  /**
   * Retrieve a subset of features for one entity.
   *
   * `HMGET` returns the requested fields in one round trip. Pass
   * `fieldNames=null` (the default) to fetch the entire hash with
   * `HGETALL` — useful for debugging but rarely the right call on the
   * request path, where the model knows exactly which features it
   * consumes.
   *
   * @param {string} entityId
   * @param {string[] | null} [fieldNames]
   * @returns {Promise<Record<string, string>>}
   */
  async getFeatures(entityId, fieldNames = null) {
    const key = this.keyFor(entityId);
    if (fieldNames === null || fieldNames === undefined) {
      const data = await this.redis.hGetAll(key);
      this.readsTotal += 1;
      this.readFieldsTotal += Object.keys(data).length;
      return data;
    }
    const names = [...fieldNames];
    if (names.length === 0) return {};
    const values = await this.redis.hmGet(key, names);
    const out = {};
    let returned = 0;
    for (let i = 0; i < names.length; i += 1) {
      const v = values[i];
      if (v !== null && v !== undefined) {
        out[names[i]] = v;
        returned += 1;
      }
    }
    this.readsTotal += 1;
    this.readFieldsTotal += returned;
    return out;
  }

  /**
   * Pipeline `HMGET` across many entities for batch scoring.
   *
   * Hundreds of entities in one round trip. The model can then score
   * them all without further network calls.
   *
   * @param {Iterable<string>} entityIds
   * @param {Iterable<string>} fieldNames
   * @returns {Promise<Record<string, Record<string, string>>>}
   */
  async batchGetFeatures(entityIds, fieldNames) {
    const ids = [...entityIds];
    const names = [...fieldNames];
    if (ids.length === 0 || names.length === 0) return {};

    const pipe = this.redis.multi();
    for (const entityId of ids) {
      pipe.hmGet(this.keyFor(entityId), names);
    }
    const rows = await pipe.exec();

    const out = {};
    let seenFields = 0;
    for (let i = 0; i < ids.length; i += 1) {
      const values = rows[i] || [];
      const row = {};
      for (let j = 0; j < names.length; j += 1) {
        const v = values[j];
        if (v !== null && v !== undefined) {
          row[names[j]] = v;
          seenFields += 1;
        }
      }
      out[ids[i]] = row;
    }
    this.readsTotal += ids.length;
    this.readFieldsTotal += seenFields;
    return out;
  }

  // --- TTL inspection (used by the demo UI) --------------------------

  /**
   * Seconds until the entity key expires.
   *
   * Returns `-1` if no key-level TTL is set, `-2` if the key doesn't
   * exist.
   *
   * @param {string} entityId
   * @returns {Promise<number>}
   */
  async keyTtlSeconds(entityId) {
    return Number(await this.redis.ttl(this.keyFor(entityId)));
  }

  /**
   * Per-field TTL via `HTTL` (Redis 7.4+).
   *
   * Each value mirrors the `TTL` convention: positive means seconds
   * remaining, `-1` means no TTL on the field, `-2` means the field
   * doesn't exist on this hash (or the key itself is missing).
   *
   * Normalized for forward-compat: some client versions can report
   * `null` for a missing key or a singleton list-of-list in pipeline
   * contexts. Both shapes collapse back to the flat list shape that
   * matches the field order passed in.
   *
   * @param {string} entityId
   * @param {Iterable<string>} fieldNames
   * @returns {Promise<Record<string, number>>}
   */
  async fieldTtlsSeconds(entityId, fieldNames) {
    const names = [...fieldNames];
    if (names.length === 0) return {};
    let ttls = await this.redis.hTTL(this.keyFor(entityId), names);
    if (ttls === null || ttls === undefined) {
      ttls = names.map(() => -2);
    } else if (Array.isArray(ttls) && ttls.length === 1 && Array.isArray(ttls[0])) {
      ttls = ttls[0];
    }
    const out = {};
    for (let i = 0; i < names.length; i += 1) {
      out[names[i]] = Number(ttls[i]);
    }
    return out;
  }

  // --- Demo housekeeping ---------------------------------------------

  /**
   * Enumerate entity IDs by scanning `keyPrefix*`.
   *
   * `SCAN` is non-blocking; the demo uses it to populate UI dropdowns,
   * not as a serving primitive.
   *
   * @param {number} [limit=200]
   * @returns {Promise<string[]>}
   */
  async listEntityIds(limit = 200) {
    const ids = [];
    const prefixLen = this.keyPrefix.length;
    for await (const key of this.redis.scanIterator({
      MATCH: `${this.keyPrefix}*`,
      COUNT: 200,
    })) {
      // node-redis 5 yields each batch as an array; older majors yield one key.
      if (Array.isArray(key)) {
        for (const k of key) {
          ids.push(k.slice(prefixLen));
          if (ids.length >= limit) break;
        }
      } else {
        ids.push(key.slice(prefixLen));
      }
      if (ids.length >= limit) break;
    }
    ids.sort();
    return ids.slice(0, limit);
  }

  /**
   * Count entities currently in the store (via `SCAN`).
   * @returns {Promise<number>}
   */
  async countEntities() {
    let count = 0;
    for await (const key of this.redis.scanIterator({
      MATCH: `${this.keyPrefix}*`,
      COUNT: 500,
    })) {
      count += Array.isArray(key) ? key.length : 1;
    }
    return count;
  }

  /**
   * @param {string} entityId
   * @returns {Promise<number>}
   */
  async deleteEntity(entityId) {
    return Number(await this.redis.del(this.keyFor(entityId)));
  }

  /**
   * Drop every entity under `keyPrefix`. Used by the demo reset path.
   *
   * Scans in batches and issues one variadic `DEL` per batch, so a
   * large demo dataset doesn't land on the server as one giant
   * synchronous delete.
   *
   * @returns {Promise<number>}
   */
  async reset() {
    let deleted = 0;
    let batch = [];
    const flush = async () => {
      if (batch.length === 0) return;
      deleted += Number(await this.redis.del(batch));
      batch = [];
    };
    for await (const key of this.redis.scanIterator({
      MATCH: `${this.keyPrefix}*`,
      COUNT: 500,
    })) {
      if (Array.isArray(key)) {
        batch.push(...key);
      } else {
        batch.push(key);
      }
      if (batch.length >= 500) await flush();
    }
    await flush();
    return deleted;
  }

  stats() {
    return {
      batch_writes_total: this.batchWritesTotal,
      streaming_writes_total: this.streamingWritesTotal,
      reads_total: this.readsTotal,
      read_fields_total: this.readFieldsTotal,
    };
  }

  resetStats() {
    this.batchWritesTotal = 0;
    this.streamingWritesTotal = 0;
    this.readsTotal = 0;
    this.readFieldsTotal = 0;
  }
}

module.exports = {
  FeatureStore,
  DEFAULT_BATCH_FIELDS,
  DEFAULT_STREAMING_FIELDS,
};
